File: src/MixNetFx/Facesso.Interfaces/Kannegiesser SQL/frmKannegiesserSQLProdDataConfigDialog.vb
Imports System.IO
Imports System.Windows.Forms
Imports ActiveDev.Data.SqlClient
Imports System.Data.SqlClient

Public Class frmKannegiesserSQLProdDataConfigDialog

    Dim mySetByAssignment As Boolean

    Private Sub btnChoosePath_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles btnChoosePath.Click
        PickSqlConnectionString()
    End Sub

    Private Sub PickSqlConnectionString()
        Dim locFB As New ADDatabaseConnectionDialog
        Dim locCB As SqlConnectionStringBuilder = locFB.GetConnectionBuilder("Kannegiesser-SQL-Verbindung:")
        If locCB IsNot Nothing Then
            txtSqlConnectionString.Text = locCB.ToString
        Else
            lblDevice.Enabled = False
            cmbDevice.Enabled = False
            Return
        End If
        
        If TryPopulateKannegiesserDevices() Then
            lblDevice.Enabled = True
            cmbDevice.Enabled = True
            DirectCast(TaskItem, KannegiesserSQLProductionDataImportTaskElement).KannegiesserSQLConnectionString = locCB.ToString
            DirectCast(TaskItem, KannegiesserSQLProductionDataImportTaskElement).KannegiesserDeviceID = Nothing
            TaskItem.ConversionItems = DirectCast(TaskItem, KannegiesserSQLProductionDataImportTaskElement).AssembleConversionItems()
            RebuildLists()
        Else
            lblDevice.Enabled = False
            cmbDevice.Enabled = False
        End If
    End Sub

    Protected Overrides Sub InitializeControls()
        MyBase.InitializeControls()
        If String.IsNullOrEmpty(DirectCast(TaskItem, KannegiesserSQLProductionDataImportTaskElement).KannegiesserSQLConnectionString) Then
            PickSqlConnectionString()
        Else
            txtSqlConnectionString.Text = DirectCast(TaskItem, KannegiesserSQLProductionDataImportTaskElement).KannegiesserSQLConnectionString
        End If
        If TryPopulateKannegiesserDevices() Then
            SetKannegiesserDevice(DirectCast(TaskItem, KannegiesserSQLProductionDataImportTaskElement).KannegiesserDeviceID)
        End If
        myAllowMultipleAssignments = True
    End Sub

    Protected Overrides ReadOnly Property BlockDeviceListBuilding() As Boolean
        Get
            Try
                If DirectCast(TaskItem, KannegiesserSQLProductionDataImportTaskElement).KannegiesserSQLConnectionString IsNot Nothing Then
                    If DirectCast(TaskItem, KannegiesserSQLProductionDataImportTaskElement).KannegiesserDeviceID IsNot Nothing Then
                        Return False
                    End If
                End If
            Catch ex As Exception
            End Try
            Return True
        End Get
    End Property

    Private Sub cmbDevice_SelectedIndexChanged(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles cmbDevice.SelectedIndexChanged
        DirectCast(TaskItem, KannegiesserSQLProductionDataImportTaskElement).KannegiesserDeviceID = GetKannegiesserDevice()
        If mySetByAssignment Then
            mysetbyAssignment = False
        Else
            TaskItem.ConversionItems = DirectCast(TaskItem, KannegiesserSQLProductionDataImportTaskElement).AssembleConversionItems()
        End If
        RebuildLists()
    End Sub

    Private Function TryPopulateKannegiesserDevices() As Boolean
        If txtSqlConnectionString.Text IsNot Nothing Then
            'Versuchen, aufzumachen
            Dim locConnection As New SqlConnection(txtSqlConnectionString.Text)
            Using locConnection
                Try
                    locConnection.Open()
                Catch ex As Exception
                    MessageBox.Show("Verbindung zur Kannegiesser-Datenbankinstanz konnte nicht hergestellt werden!", _
                                    "Verbindungsfehler!", MessageBoxButtons.OK, MessageBoxIcon.Exclamation)
                    Return False
                End Try

                Dim oc As New KannegiesserDataContext(txtSqlConnectionString.Text)
                Dim machineList = (From item In oc.GetMachines).ToList
                cmbDevice.Items.Clear()
                cmbDevice.Items.Add("- nicht definiert -")
                If machineList.Count > 0 Then
                    For Each item In machineList
                        cmbDevice.Items.Add(item)
                    Next
                End If
            End Using
            Return True
        End If
    End Function

    Private Function GetKannegiesserDevice() As String
        If cmbDevice.SelectedIndex <= 0 Then
            Return Nothing
        End If

        If cmbDevice.SelectedItem IsNot Nothing Then
            Return DirectCast(cmbDevice.SelectedItem, GetMachinesResult).MachineID.ToString
        End If
        Return Nothing
    End Function

    Private Sub SetKannegiesserDevice(ByVal Device As String)
        mySetByAssignment = True
        If String.IsNullOrEmpty(Device) Then
            cmbDevice.SelectedIndex = 0
            Return
        End If

        For locIndex As Integer = 0 To cmbDevice.Items.Count - 1
            If locIndex > 0 Then
                If DirectCast(cmbDevice.Items(locIndex), GetMachinesResult).MachineID.ToString = Device Then
                    cmbDevice.SelectedIndex = locIndex
                    Return
                End If
            End If
        Next
        cmbDevice.SelectedIndex = 0
    End Sub
End Class

File: src/NetFxCSharp/Facesso.Interfaces/Kannegiesser SQL/frmKannegiesserSQLProdDataConfigDialog.cs
using ActiveDev.Data.SqlClient;
using Microsoft.VisualBasic;
using System;
using System.Collections;
using System.Collections.Generic;
using System.Data;
using System.Data.SqlClient;
using System.Diagnostics;
using System.IO;
using System.Linq;
using System.Threading.Tasks;
using System.Windows.Forms;
using System.Xml.Linq;

namespace Facesso.Interfaces
{
    public partial class frmKannegiesserSQLProdDataConfigDialog
    {
        private bool mySetByAssignment;
        private void btnChoosePath_Click(System.Object sender, System.EventArgs e)
        {
            PickSqlConnectionString();
        }

        private void PickSqlConnectionString()
        {
            ADDatabaseConnectionDialog locFB = new ADDatabaseConnectionDialog();
            SqlConnectionStringBuilder locCB = locFB.GetConnectionBuilder("Kannegiesser-SQL-Verbindung:");
            if (locCB != null)
            {
                txtSqlConnectionString.Text = locCB.ToString();
            }
            else
            {
                lblDevice.Enabled = false;
                cmbDevice.Enabled = false;
                return;
            }

            if (TryPopulateKannegiesserDevices())
            {
                lblDevice.Enabled = true;
                cmbDevice.Enabled = true;
                ((KannegiesserSQLProductionDataImportTaskElement)TaskItem).KannegiesserSQLConnectionString = locCB.ToString();
                ((KannegiesserSQLProductionDataImportTaskElement)TaskItem).KannegiesserDeviceID = null;
                TaskItem.ConversionItems = ((KannegiesserSQLProductionDataImportTaskElement)TaskItem).AssembleConversionItems();
                RebuildLists();
            }
            else
            {
                lblDevice.Enabled = false;
                cmbDevice.Enabled = false;
            }
        }

        protected override void InitializeControls()
        {
            base.InitializeControls();
            if (string.IsNullOrEmpty(((KannegiesserSQLProductionDataImportTaskElement)TaskItem).KannegiesserSQLConnectionString))
            {
                PickSqlConnectionString();
            }
            else
            {
                txtSqlConnectionString.Text = ((KannegiesserSQLProductionDataImportTaskElement)TaskItem).KannegiesserSQLConnectionString;
            }

            if (TryPopulateKannegiesserDevices())
            {
                SetKannegiesserDevice(((KannegiesserSQLProductionDataImportTaskElement)TaskItem).KannegiesserDeviceID);
            }

            myAllowMultipleAssignments = true;
        }

        protected override bool BlockDeviceListBuilding
        {
            get
            {
                try
                {
                    if (((KannegiesserSQLProductionDataImportTaskElement)TaskItem).KannegiesserSQLConnectionString != null)
                    {
                        if (((KannegiesserSQLProductionDataImportTaskElement)TaskItem).KannegiesserDeviceID != null)
                        {
                            return false;
                        }
                    }
                }
                catch (Exception ex)
                {
                }

                return true;
            }
        }

        private void cmbDevice_SelectedIndexChanged(System.Object sender, System.EventArgs e)
        {
            ((KannegiesserSQLProductionDataImportTaskElement)TaskItem).KannegiesserDeviceID = GetKannegiesserDevice();
            if (mySetByAssignment)
            {
                mySetByAssignment = false;
            }
            else
            {
                TaskItem.ConversionItems = ((KannegiesserSQLProductionDataImportTaskElement)TaskItem).AssembleConversionItems();
            }

            RebuildLists();
        }

        private bool TryPopulateKannegiesserDevices()
        {
            if (txtSqlConnectionString.Text != null)
            {
                //Versuchen, aufzumachen
                SqlConnection locConnection = new SqlConnection(txtSqlConnectionString.Text);
                using (locConnection)
                {
                    try
                    {
                        locConnection.Open();
                    }
                    catch (Exception ex)
                    {
                        MessageBox.Show("Verbindung zur Kannegiesser-Datenbankinstanz konnte nicht hergestellt werden!", "Verbindungsfehler!", MessageBoxButtons.OK, MessageBoxIcon.Exclamation);
                        return false;
                    }

                    KannegiesserDataContext oc = new KannegiesserDataContext(txtSqlConnectionString.Text);
                    var machineList = ((
                        from item in oc.GetMachines()select item)).ToList();
                    cmbDevice.Items.Clear();
                    cmbDevice.Items.Add("- nicht definiert -");
                    if (machineList.Count > 0)
                    {
                        foreach (var item in machineList)
                        {
                            cmbDevice.Items.Add(item);
                        }
                    }
                }

                return true;
            }

            return default(bool);
        }

        private string GetKannegiesserDevice()
        {
            if (cmbDevice.SelectedIndex <= 0)
            {
                return null;
            }

            if (cmbDevice.SelectedItem != null)
            {
                return ((GetMachinesResult)cmbDevice.SelectedItem).MachineID.ToString();
            }

            return null;
        }

        private void SetKannegiesserDevice(string Device)
        {
            mySetByAssignment = true;
            if (string.IsNullOrEmpty(Device))
            {
                cmbDevice.SelectedIndex = 0;
                return;
            }

            for (int locIndex = 0; locIndex <= cmbDevice.Items.Count - 1; locIndex++)
            {
                if (locIndex > 0)
                {
                    if (((GetMachinesResult)cmbDevice.Items[locIndex]).MachineID.ToString() == Device)
                    {
                        cmbDevice.SelectedIndex = locIndex;
                        return;
                    }
                }
            }

            cmbDevice.SelectedIndex = 0;
        }

        public frmKannegiesserSQLProdDataConfigDialog()
        {
            InitializeComponent();
        }
    }
}